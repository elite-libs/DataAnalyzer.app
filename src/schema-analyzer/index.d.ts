// export type TypeDescriptorName = 'enum' | 'nullable' | 'unique';
// export type TypeNameString = '$ref' |
//     'Unknown' |
//     'ObjectId' |
//     'UUID' |
//     'Boolean' |
//     'Date' |
//     'Timestamp' |
//     'Currency' |
//     'Float' |
//     'Number' |
//     'Email' |
//     'String' |
//     'Array' |
//     'Object' |
//     'Null'

// export interface ISchemaAnalyzerOptions {
//     onProgress?: progressCallback | undefined;
//     enumMinimumRowCount?: number | undefined;
//     enumAbsoluteLimit?: number | undefined;
//     enumPercentThreshold?: number | undefined;
//     nullableRowsThreshold?: number | undefined;
//     uniqueRowsThreshold?: number | undefined;
//     strictMatching?: boolean | undefined;
//     disableNestedTypes?: boolean | undefined;
//     bogusSizeThreshold?: number | undefined,
// }

// /**
//  * Includes the results of main top-level schema.
//  */
// export type TypeSummary = {
//     fields: {
//         [x: string]: FieldInfo;
//     };
//     totalRows: number;
//     nestedTypes?: {
//         [x: string]: TypeSummary;
//     } | undefined;
// };
// /**
//  * Describes one or more potential types discovered for a field. The `types` object will have a `$ref` key if any nested structures were found.
//  */
// export type FieldInfo = {
//     /**
//      * - field stats organized by type
//      */
//     types: {
//         [x: TypeNameString]: FieldTypeSummary;
//     };
//     /**
//      * - is the field nullable
//      */
//     nullable: boolean;
//     /**
//      * - is the field unique
//      */
//     unique: boolean;
//     /**
//      * - enumeration detected, the values are listed on this property.
//      */
//     enum?: string[] | number[] | undefined;
// };
// /**
//  * Contains stats for a given field's (potential) type.
//  *
//  * TODO: Add string property for the type name.
//  *    We currently uses object key structure: {"String": FieldTypeSummary}
//  */
// export type FieldTypeSummary = {
//     /**
//      * - for nested type support.
//      */
//     typeAlias?: string | undefined;
//     /**
//      * - extracted field values, placed into an array. This simplifies (at expense of memory) type analysis and summarization when creating the `AggregateSummary`.
//      */
//     value?: AggregateSummary | undefined;
//     /**
//      * - summary of array of string (or decimal) sizes, pre processing into an AggregateSummary
//      */
//     length?: AggregateSummary | undefined;
//     /**
//      * - only applies to Float types. Summary of array of sizes of the value both before and after the decimal.
//      */
//     precision?: AggregateSummary | undefined;
//     /**
//      * - only applies to Float types. Summary of array of sizes of the value after the decimal.
//      */
//     scale?: AggregateSummary | undefined;
//     /**
//      * - if enum rules were triggered will contain the detected unique values.
//      */
//     enum?: string[] | number[] | undefined;
//     /**
//      * - number of times the type was matched
//      */
//     count: number;
//     /**
//      * - absolute priority of the detected TypeName, defined in the object `typeRankings`
//      */
//     rank: number;
// };
// /**
//  * This is an internal intermediate structure.
//  * It mirrors the `FieldSummary` type it will become.
//  */
// export type InternalFieldTypeData = {
//     /**
//      * - array of values, pre processing into an AggregateSummary
//      */
//     value?: any[] | undefined;
//     /**
//      * - array of string (or decimal) sizes, pre processing into an AggregateSummary
//      */
//     length?: number[] | undefined;
//     /**
//      * - only applies to Float types. Array of sizes of the value both before and after the decimal.
//      */
//     precision?: number[] | undefined;
//     /**
//      * - only applies to Float types. Array of sizes of the value after the decimal.
//      */
//     scale?: number[] | undefined;
//     /**
//      * - number of times the type was matched
//      */
//     count?: number | undefined;
//     /**
//      * - absolute priority of the detected TypeName, defined in the object `typeRankings`
//      */
//     rank?: number | undefined;
// };
// /**
//  * Used to represent a number series of any size.
//  * Includes the lowest (`min`), highest (`max`), mean/average (`mean`) and measurements at certain `percentiles`.
//  */
// export type AggregateSummary = {
//     min: number;
//     max: number;
//     mean: number;
//     p25: number;
//     p33: number;
//     p50: number;
//     p66: number;
//     p75: number;
//     p99: number;
// };
// /**
//  * This callback is displayed as a global member.
//  */
// export type progressCallback = (progress: {
//     totalRows: number;
//     currentRow: number;
// }) => any;
// /**
//  * Includes the results of main top-level schema.
//  * @typedef TypeSummary
//  * @type {{
//  *  fields: Object.<string, FieldInfo>;
//  *  totalRows: number;
//  *  nestedTypes?: Object.<string, TypeSummary>;
//  * }}
//  */
// /**
//  * Describes one or more potential types discovered for a field. The `types` object will have a `$ref` key if any nested structures were found.
//  * @typedef FieldInfo
//  * @type {object}
//  * @property {Object.<string, string | FieldTypeSummary>} types - field stats organized by type
//  * @property {boolean} nullable - is the field nullable
//  * @property {boolean} unique - is the field unique
//  * @property {string[]|number[]} [enum] - enumeration detected, the values are listed on this property.
//  */
// /**
//  * Contains stats for a given field's (potential) type.
//  *
//  * TODO: Add string property for the type name.
//  *    We currently uses object key structure: {"String": FieldTypeSummary}
//  * @typedef FieldTypeSummary
//  * @type {Object}
//  * @property {string} [typeAlias] - for nested type support.
//  * @property {AggregateSummary} [value] - extracted field values, placed into an array. This simplifies (at expense of memory) type analysis and summarization when creating the `AggregateSummary`.
//  * @property {AggregateSummary} [length] - summary of array of string (or decimal) sizes, pre processing into an AggregateSummary
//  * @property {AggregateSummary} [precision] - only applies to Float types. Summary of array of sizes of the value both before and after the decimal.
//  * @property {AggregateSummary} [scale] - only applies to Float types. Summary of array of sizes of the value after the decimal.
//  * @property {string[]|number[]} [enum] - if enum rules were triggered will contain the detected unique values.
//  * @property {number} count - number of times the type was matched
//  * @property {number} rank - absolute priority of the detected TypeName, defined in the object `typeRankings`
//  */
// /**
//  * This is an internal intermediate structure.
//  * It mirrors the `FieldSummary` type it will become.
//  * @private
//  * @typedef InternalFieldTypeData
//  * @type {Object}
//  * @property {any[]} [value] - array of values, pre processing into an AggregateSummary
//  * @property {number[]} [length] - array of string (or decimal) sizes, pre processing into an AggregateSummary
//  * @property {number[]} [precision] - only applies to Float types. Array of sizes of the value both before and after the decimal.
//  * @property {number[]} [scale] - only applies to Float types. Array of sizes of the value after the decimal.
//  * @property {number} [count] - number of times the type was matched
//  * @property {number} [rank] - absolute priority of the detected TypeName, defined in the object `typeRankings`
//  *
//  */
// /**
//  * Used to represent a number series of any size.
//  * Includes the lowest (`min`), highest (`max`), mean/average (`mean`) and measurements at certain `percentiles`.
//  * @typedef AggregateSummary
//  * @type {{min: number, max: number, mean: number, p25: number, p33: number, p50: number, p66: number, p75: number, p99: number}}
//  */
// /**
//  * This callback is displayed as a global member.
//  * @callback progressCallback
//  * @param {{totalRows: number, currentRow: number}} progress - The current progress of processing.
//  */
// /**
//  * schemaAnalyzer() is the main function and where all the analysis & processing happens.
//  * @param {string} schemaName The name, or name prefix to use when assembling results. Helpful with nested types (aka sub-types.)
//  * @param {Array<Object>} input - The input data to analyze. Must be an array of objects.
//  * @param {ISchemaAnalyzerOptions} [options] - Optional parameters
//  * @returns {Promise<TypeSummary>} Returns and
//  */
// export function schemaAnalyzer(schemaName: string, input: Array<Object>, options?: ISchemaAnalyzerOptions | undefined): Promise<TypeSummary>;

// export function pivotFieldDataByType(typeSizeData: any): any;
// /**
//  * Accepts an array of numbers and returns summary data about
//  *  the range & spread of points in the set.
//  *
//  * @param {number[]} numbers - sequence of unsorted data points
//  * @returns {AggregateSummary}
//  */
// export function getNumberRangeStats(numbers: number[], useSortedDataForPercentiles?: boolean): AggregateSummary;
// export function isValidDate(date: any): any;
// /**
//  * Returns a fieldName keyed-object with type detection summary data.
//  *
//  * ### Example `fieldSummary`:
//  * ```
//  * {
//  *  "id": {
//  *    "UUID": {
//  *      "rank": 2,
//  *      "count": 25
//  *    },
//  *    "Number": {
//  *      "rank": 8,
//  *      "count": 1,
//  *      "value": {
//  *        "min": 9999999,
//  *        "mean": 9999999,
//  *        "max": 9999999,
//  *        "p25": 9999999,
//  *        "p33": 9999999,
//  *        "p50": 9999999,
//  *        "p66": 9999999,
//  *        "p75": 9999999,
//  *        "p99": 9999999
//  *      }
//  *    }
//  *  }
//  * }
//  * ```
//  *
//  * @param {{ fieldsData: Object.<string, InternalFieldTypeData[]>, uniques: Object.<string, any[]>, totalRows: number}} schema
//  * @returns {{
//  *  fields: Object.<string, Object.<string, FieldTypeSummary>>,
//  *  uniques: Object.<string, any[]>,
//  *  totalRows: number
//  * }}
//  */
// declare function condenseFieldData(schema: {
//     fieldsData: {
//         [x: string]: InternalFieldTypeData[];
//     };
//     uniques: {
//         [x: string]: any[];
//     };
//     totalRows: number;
// }): {
//     fields: {
//         [x: string]: {
//             [x: TypeNameString]: FieldTypeSummary;
//         };
//     };
//     uniques: {
//         [x: string]: any[];
//     };
//     totalRows: number;
// };
// /**
//  *
//  */
// declare function formatRangeStats(stats: any, formatter: any): {
//     min: any;
//     mean: any;
//     max: any;
//     p25: any;
//     p33: any;
//     p50: any;
//     p66: any;
//     p75: any;
//     p99: any;
// };
// export { condenseFieldData as _condenseFieldData, pivotFieldDataByType as _pivotFieldDataByType, getNumberRangeStats as _getNumberRangeStats, formatRangeStats as _formatRangeStats };
