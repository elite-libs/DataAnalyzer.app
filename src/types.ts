export interface Dictionary<T> {
  [id: string]: T | undefined | null;
  [id: number]: T | undefined | null;
}

export type CallbackFn<TArgs, TReturn = void> = (args?: TArgs | any) => TReturn | any;

export interface Dict<TValue> {
  [id: string]: TValue | undefined | null;
}
