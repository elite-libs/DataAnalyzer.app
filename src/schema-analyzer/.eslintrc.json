{
  "parserOptions": {
    "ecmaVersion": 2020,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "env": {
    "es6": true,
    "node": true,
    "browser": true,
    "jest": true
  },
  "plugins": ["import", "node", "promise", "prettier"],
  "extends": [
    "prettier",
    "prettier/@typescript-eslint",
    "prettier/babel",
    "prettier/react"
  ]
}
