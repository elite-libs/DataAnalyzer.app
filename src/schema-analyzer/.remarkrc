{
  "settings": {
    "listItemIndent": "mixed"
  },
  "plugins": [
    "remark-stringify",
    ["remark-lint-list-item-indent", false]
  ]
}