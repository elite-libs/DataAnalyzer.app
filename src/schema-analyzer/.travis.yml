language: node_js
node_js:
    - 12
install:
    - npm install -g codecov
script:
    - npm ci
    - npm test
    - codecov
