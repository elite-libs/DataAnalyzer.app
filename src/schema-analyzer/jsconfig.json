{
  "compilerOptions": {
    "jsx": "react",
    "checkJs": true,
    "resolveJsonModule": true,
    "esModuleInterop": true
  },
  "typeAcquisition": {
    "include": [
      "jest"
    ]
  }
}