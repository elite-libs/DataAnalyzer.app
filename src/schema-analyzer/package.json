{
  "name": "schema-analyzer",
  "version": "2.0.0",
  "description": "Heuristic \"column\" type & size analysis w/ enumeration detection.",
  "type": "commonjs",
  "module": "index.tsx",
  "main": "index.tsx",
  "browser": "dist/schema-analyzer.bundle.js",
  "files": [
    "dist",
    "build/index.d.ts",
    "index.ts"
  ],
  "scripts": {
    "prebuild": "rm -rf dist/*",
    "build": "npm run build:webpack && NODE_ENV=production rollup -c && NODE_ENV=development rollup -c && npm run compress",
    "build:webpack": "webpack -c webpack.config.js",
    "coverage": "jest --coverage --maxWorkers 4 && codecov -f coverage/coverage-final.json",
    "postcoverage": "codecov --token=$CODECOV_TOKEN",
    "dev": "rollup -c -w",
    "test": "jest --runTestsByPath --coverage --maxWorkers 4 ./index.test.ts ./utils/type-helpers.test.ts",
    "test:debug": "node --inspect-brk=9823 ./node_modules/.bin/jest --runTestsByPath ./index.test.ts ./utils/type-*.test.ts",
    "test:watch": "jest --watch",
    "pretest": "npm run build",
    "format": "prettier -w *.js index*.ts utils/*.ts",
    "compress": "zopfli -i2000 --gzip ./dist/schema-analyzer*min.js && brotli --keep --quality=11 -- ./dist/schema-analyzer*min.js"
  },
  "engines": {
    "node": "< 14"
  },
  "types": "./index.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/justsml/schema-analyzer.git"
  },
  "keywords": [],
  "author": {
    "name": "Dan Levy",
    "url": "https://danlevy.net/"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/justsml/schema-analyzer/issues"
  },
  "homepage": "https://github.com/justsml/schema-analyzer#readme",
  "dependencies": {
    "debug": "^4.3.1",
    "ts-node": "^9.1.1"
  },
  "devDependencies": {
    "@babel/core": "~7.8.3",
    "@babel/helper-compilation-targets": "~7.12.5",
    "@babel/plugin-transform-runtime": "~7.8.3",
    "@babel/preset-env": "~7.8.3",
    "@babel/preset-typescript": "^7.12.7",
    "@babel/runtime-corejs3": "~7.8.3",
    "@rollup/plugin-commonjs": "~11.0.1",
    "@rollup/plugin-node-resolve": "~7.0.0",
    "@rollup/plugin-typescript": "8.1.0",
    "@types/debug": "^4.1.5",
    "@types/jest": "~25.1.0",
    "@types/lodash-es": "^4.17.4",
    "babel-eslint": "~10.0.3",
    "babel-jest": "26.6.3",
    "codecov": "3.8.1",
    "core-js": "3.8.2",
    "csv-parse": "~4.8.5",
    "eslint-config-standard": "~14.1.0",
    "eslint-config-standard-react": "~9.2.0",
    "eslint-plugin-import": "2.22.1",
    "eslint-plugin-node": "~11.0.0",
    "eslint-plugin-promise": "~4.2.1",
    "eslint-plugin-react": "7.22.0",
    "eslint-plugin-standard": "~4.0.1",
    "jest": "26.6.3",
    "lodash-es": "^4.17.20",
    "prettier": "^2.2.1",
    "rollup": "~1.30.1",
    "rollup-plugin-node-builtins": "^2.1.2",
    "rollup-plugin-node-globals": "^1.4.0",
    "rollup-plugin-terser": "~5.2.0",
    "standard": "16.0.3",
    "terser-webpack-plugin": "^4.2.3",
    "ts-jest": "^26.4.4",
    "ts-loader": "^8.0.14",
    "tslib": "^2.1.0",
    "typescript": "^4.1.3",
    "webpack": "^5.17.0",
    "webpack-cli": "^4.4.0"
  },
  "prettier": {
    "endOfLine": "lf",
    "trailingComma": "all",
    "tabWidth": 2,
    "semi": false,
    "singleQuote": true
  },
  "jest": {
    "moduleFileExtensions": [
      "ts",
      "js",
      "mjs"
    ],
    "preset": "ts-jest",
    "moduleNameMapper": {
      "^lodash-es$": "<rootDir>/node_modules/lodash/index.js"
    }
  },
  "babel": {
    "plugins": [],
    "presets": [
      [
        "@babel/preset-env",
        {
          "corejs": {
            "version": "3.8.2",
            "proposals": true
          },
          "modules": "auto",
          "useBuiltIns": "usage",
          "targets": {
            "node": 14,
            "browsers": [
              "Chrome >= 60",
              "Safari >= 10.1",
              "iOS >= 10.3",
              "Firefox >= 54",
              "Edge >= 15"
            ]
          }
        }
      ],
      "@babel/preset-typescript"
    ]
  },
  "standard": {
    "env": [
      "jest"
    ]
  },
  "eslintIgnore": [
    "node_modules",
    "dist"
  ],
  "eslintConfig": {
    "parserOptions": {
      "ecmaVersion": 2020,
      "ecmaFeatures": {
        "jsx": true
      }
    },
    "env": {
      "es6": true,
      "node": true,
      "browser": true,
      "jest": true
    },
    "plugins": [
      "import",
      "node",
      "promise",
      "prettier"
    ]
  },
  "husky": {
    "hooks": {
      "pre-commit": "npm run lint",
      "pre-push": "CI=true npm test"
    }
  },
  "browserslist": [
    ">2%",
    "not dead",
    "not ie <= 11",
    "not op_mini all"
  ],
  "remarkConfig": {
    "settings": {
      "listItemIndent": "mixed"
    },
    "plugins": [
      "remark-stringify",
      [
        "remark-lint-list-item-indent",
        false
      ]
    ]
  }
}
