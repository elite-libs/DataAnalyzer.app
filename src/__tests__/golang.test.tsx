import golang from '../adapters/writer.golang';
import { schemaAnalyzer } from '../schema-analyzer/index';
import usersNotes from '../../public/data/user-notes.json';
import users from '../../public/data/users.example.json';
import pokemonCharmander from '../../public/data/pokemon-charmander.json';
import { ISchemaAnalyzerOptions } from 'types';

describe('#GoLang', () => {
  it('can emit structs with enum', async () => {
    const options: ISchemaAnalyzerOptions = {
      strictMatching: false,
      flattenOptions: {
        targetLength: 'p99',
        targetPrecision: 'p99',
        targetScale: 'p99',
        targetValue: 'p99',
        nullableRowsThreshold: 0.001,
      },
    };
    const results = await schemaAnalyzer('users', users, options);
    const code = golang.render(results);

    expect(code).toMatchSnapshot();
  });

  it('can emit nested structs', async () => {
    const options: ISchemaAnalyzerOptions = {
      strictMatching: false,
      flattenOptions: {
        targetLength: 'max',
        targetPrecision: 'max',
        targetScale: 'max',
        targetValue: 'max',
        nullableRowsThreshold: 0.001,
      },
    };
    const results = await schemaAnalyzer('User', usersNotes, options);
    const code = golang.render(results);

    expect(code).toMatchSnapshot();
  });

  it('can emit deeply nested structs', async () => {
    const options: ISchemaAnalyzerOptions = {
      strictMatching: false,
      flattenOptions: {
        targetLength: 'max',
        targetPrecision: 'max',
        targetScale: 'max',
        targetValue: 'max',
        nullableRowsThreshold: 0.001,
      },
    };
    const results = await schemaAnalyzer('Pokemon', pokemonCharmander, options);
    const code = golang.render(results);

    expect(code).toMatchSnapshot();
  });

  it('can emit consolidated nested structs', async () => {
    const options: ISchemaAnalyzerOptions = {
      strictMatching: false,
      consolidateTypes: 'field-names',
      debug: true,
      flattenOptions: {
        targetLength: 'max',
        targetPrecision: 'max',
        targetScale: 'max',
        targetValue: 'max',
        nullableRowsThreshold: 0.001,
      },
    };
    const results = await schemaAnalyzer('Pokemon', pokemonCharmander, options);
    const code = golang.render(results);

    expect(code).toMatchSnapshot();
  });

  // it('can emit golang', async () => {
  //   const options = { strictMatching: false };
  //   const results = await schemaAnalyzer('users', usersSparse, options);

  //   expect(results.fields.id).toBeDefined();
  //   expect(results.fields.id?.unique).toBe(true);
  //   expect(results.fields.id?.identity).toBe(true);
  //   expect(results.fields.id?.nullable).toBe(false);
  // });
});

// golang
