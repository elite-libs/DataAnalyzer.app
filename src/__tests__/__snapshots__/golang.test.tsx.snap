// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#GoLang can emit deeply nested structs 1`] = `
"
type Pokemon struct {
    Abilities []PokemonAbilities
    BaseExperience int
    Forms []PokemonForms
    GameIndices []PokemonGameIndices
    Height int
    HeldItems []string
    Id int
    IsDefault bool
    LocationAreaEncounters string
    Moves []PokemonMoves
    Name string
    Order int
    Species PokemonSpecies
    Sprites PokemonSprites
    Stats []PokemonStats
    Types []PokemonTypes
    Weight int
}

type PokemonAbilities struct {
    Ability PokemonAbilitiesAbility
    IsHidden bool
    Slot int
}

type PokemonForms struct {
    Name string
    Url string
}

type PokemonGameIndices struct {
    GameIndex int
    Version PokemonGameIndicesVersion
}

type PokemonMoves struct {
    Move PokemonMovesMove
    VersionGroupDetails []PokemonMovesVersionGroupDetails
}

type PokemonSpecies struct {
    Name string
    Url string
}

type PokemonSprites struct {
    BackDefault string
    BackFemale string
    BackShiny string
    BackShinyFemale string
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
    Other PokemonSpritesOther
    Versions PokemonSpritesVersions
}

type PokemonStats struct {
    BaseStat int
    Effort int
    Stat PokemonStatsStat
}

type PokemonTypes struct {
    Slot int
    Type PokemonTypesType
}

type PokemonAbilitiesAbility struct {
    Name string
    Url string
}

type PokemonGameIndicesVersion struct {
    Name string
    Url string
}

type PokemonMovesMove struct {
    Name string
    Url string
}

type PokemonMovesVersionGroupDetails struct {
    LevelLearnedAt int
    MoveLearnMethod PokemonMovesVersionGroupDetailsMoveLearnMethod
    VersionGroup PokemonMovesVersionGroupDetailsVersionGroup
}

type PokemonMovesVersionGroupDetailsMoveLearnMethod struct {
    Name NameEnum
    Url UrlEnum
}

type PokemonMovesVersionGroupDetailsVersionGroup struct {
    Name string
    Url string
}

type PokemonSpritesOther struct {
    DreamWorld PokemonSpritesOtherDreamWorld
    OfficialArtwork PokemonSpritesOtherOfficialArtwork
}

type PokemonSpritesVersions struct {
    GenerationI PokemonSpritesVersionsGenerationI
    GenerationIi PokemonSpritesVersionsGenerationIi
    GenerationIii PokemonSpritesVersionsGenerationIii
    GenerationIv PokemonSpritesVersionsGenerationIv
    GenerationV PokemonSpritesVersionsGenerationV
    GenerationVi PokemonSpritesVersionsGenerationVi
    GenerationVii PokemonSpritesVersionsGenerationVii
    GenerationViii PokemonSpritesVersionsGenerationViii
}

type PokemonSpritesOtherDreamWorld struct {
    FrontDefault string
    FrontFemale string
}

type PokemonSpritesOtherOfficialArtwork struct {
    FrontDefault string
}

type PokemonSpritesVersionsGenerationI struct {
    RedBlue PokemonSpritesVersionsGenerationIRedBlue
    Yellow PokemonSpritesVersionsGenerationIYellow
}

type PokemonSpritesVersionsGenerationIi struct {
    Crystal PokemonSpritesVersionsGenerationIiCrystal
    Gold PokemonSpritesVersionsGenerationIiGold
    Silver PokemonSpritesVersionsGenerationIiSilver
}

type PokemonSpritesVersionsGenerationIii struct {
    Emerald PokemonSpritesVersionsGenerationIiiEmerald
    FireredLeafgreen PokemonSpritesVersionsGenerationIiiFireredLeafgreen
    RubySapphire PokemonSpritesVersionsGenerationIiiRubySapphire
}

type PokemonSpritesVersionsGenerationIv struct {
    DiamondPearl PokemonSpritesVersionsGenerationIvDiamondPearl
    HeartgoldSoulsilver PokemonSpritesVersionsGenerationIvHeartgoldSoulsilver
    Platinum PokemonSpritesVersionsGenerationIvPlatinum
}

type PokemonSpritesVersionsGenerationV struct {
    BlackWhite PokemonSpritesVersionsGenerationVBlackWhite
}

type PokemonSpritesVersionsGenerationVi struct {
    OmegarubyAlphasapphire PokemonSpritesVersionsGenerationViOmegarubyAlphasapphire
    XY PokemonSpritesVersionsGenerationViXY
}

type PokemonSpritesVersionsGenerationVii struct {
    Icons PokemonSpritesVersionsGenerationViiIcons
    UltraSunUltraMoon PokemonSpritesVersionsGenerationViiUltraSunUltraMoon
}

type PokemonSpritesVersionsGenerationViii struct {
    Icons PokemonSpritesVersionsGenerationViiiIcons
}

type PokemonSpritesVersionsGenerationIRedBlue struct {
    BackDefault string
    BackGray string
    FrontDefault string
    FrontGray string
}

type PokemonSpritesVersionsGenerationIYellow struct {
    BackDefault string
    BackGray string
    FrontDefault string
    FrontGray string
}

type PokemonSpritesVersionsGenerationIiCrystal struct {
    BackDefault string
    BackShiny string
    FrontDefault string
    FrontShiny string
}

type PokemonSpritesVersionsGenerationIiGold struct {
    BackDefault string
    BackShiny string
    FrontDefault string
    FrontShiny string
}

type PokemonSpritesVersionsGenerationIiSilver struct {
    BackDefault string
    BackShiny string
    FrontDefault string
    FrontShiny string
}

type PokemonSpritesVersionsGenerationIiiEmerald struct {
    FrontDefault string
    FrontShiny string
}

type PokemonSpritesVersionsGenerationIiiFireredLeafgreen struct {
    BackDefault string
    BackShiny string
    FrontDefault string
    FrontShiny string
}

type PokemonSpritesVersionsGenerationIiiRubySapphire struct {
    BackDefault string
    BackShiny string
    FrontDefault string
    FrontShiny string
}

type PokemonSpritesVersionsGenerationIvDiamondPearl struct {
    BackDefault string
    BackFemale string
    BackShiny string
    BackShinyFemale string
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationIvHeartgoldSoulsilver struct {
    BackDefault string
    BackFemale string
    BackShiny string
    BackShinyFemale string
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationIvPlatinum struct {
    BackDefault string
    BackFemale string
    BackShiny string
    BackShinyFemale string
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationVBlackWhite struct {
    Animated PokemonSpritesVersionsGenerationVBlackWhiteAnimated
    BackDefault string
    BackFemale string
    BackShiny string
    BackShinyFemale string
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationVBlackWhiteAnimated struct {
    BackDefault string
    BackFemale string
    BackShiny string
    BackShinyFemale string
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationViOmegarubyAlphasapphire struct {
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationViXY struct {
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationViiIcons struct {
    FrontDefault string
    FrontFemale string
}

type PokemonSpritesVersionsGenerationViiUltraSunUltraMoon struct {
    FrontDefault string
    FrontFemale string
    FrontShiny string
    FrontShinyFemale string
}

type PokemonSpritesVersionsGenerationViiiIcons struct {
    FrontDefault string
    FrontFemale string
}

type PokemonStatsStat struct {
    Name string
    Url string
}

type PokemonTypesType struct {
    Name string
    Url string
}
"
`;

exports[`#GoLang can emit nested structs 1`] = `
"
type Users struct {
    Id int
    Name string
    Notes []UserNotes
}

type UserNotes struct {
    Id int
    Text string
}
"
`;

exports[`#GoLang can emit structs with enum 1`] = `
"
type Users struct {
    Id UUID
    Email string
    FirstName string
    LastName string
    IsActive bool
    CreatedAt Time
    ModifiedAt Time
    CatchPhrase string
    CompanyName string
    Credits CreditsEnum
    SalesScore float32
    Location LocationEnum
    Profile string
}
"
`;

exports[`#GoLang can emit structs with enum 2`] = `
Object {
  "fields": Object {
    "catchPhrase": Object {
      "count": 26,
      "enum": null,
      "identity": false,
      "length": 48,
      "nullable": false,
      "type": "String",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "companyName": Object {
      "count": 26,
      "enum": null,
      "identity": false,
      "length": 20,
      "nullable": false,
      "type": "String",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "createdAt": Object {
      "count": 26,
      "enum": null,
      "identity": false,
      "nullable": false,
      "type": "Date",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "credits": Object {
      "count": 26,
      "enum": Array [
        0,
        1,
        10,
        11,
        15,
        20,
        7,
        9,
      ],
      "identity": false,
      "nullable": false,
      "type": "Number",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "email": Object {
      "count": 25,
      "enum": null,
      "identity": false,
      "length": 29,
      "nullable": false,
      "type": "String",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "firstName": Object {
      "count": 25,
      "enum": null,
      "identity": false,
      "length": 10,
      "nullable": false,
      "type": "String",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "id": Object {
      "count": 25,
      "enum": null,
      "identity": true,
      "nullable": false,
      "type": "UUID",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": true,
    },
    "isActive": Object {
      "count": 26,
      "enum": null,
      "identity": false,
      "nullable": false,
      "type": "Boolean",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "lastName": Object {
      "count": 25,
      "enum": null,
      "identity": false,
      "length": 11,
      "nullable": false,
      "type": "String",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "location": Object {
      "count": 26,
      "enum": Array [
        "denver",
        "nyc",
        "remote",
      ],
      "identity": false,
      "length": 6,
      "nullable": false,
      "type": "String",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "modifiedAt": Object {
      "count": 26,
      "enum": null,
      "identity": false,
      "nullable": false,
      "type": "Date",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
    "profile": Object {
      "count": 25,
      "enum": null,
      "identity": false,
      "length": 234,
      "nullable": true,
      "type": "String",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": true,
    },
    "salesScore": Object {
      "count": 26,
      "enum": null,
      "identity": false,
      "nullable": false,
      "precision": 9,
      "scale": 7,
      "type": "Float",
      "typeRef": undefined,
      "typeRelationship": undefined,
      "unique": false,
    },
  },
  "nestedTypes": Object {},
  "schemaName": "users",
  "totalRows": 26,
}
`;
